File: core/src/filesystem.rs
use crate::errors::{CoreError, CoreResult};
use crate::models::FileEntry;
use crate::security::{build_exclude_set, is_excluded, WorkspaceGuard};
use base64::Engine;
use serde::Deserialize;
use serde_json::{json, Value};
use std::fs;
use std::io::Write;
use std::path::Path;
use std::time::UNIX_EPOCH;

const DEFAULT_MAX_READ_BYTES: u64 = 10 * 1024 * 1024;
const DEFAULT_MAX_WRITE_BYTES: u64 = 10 * 1024 * 1024;
const DEFAULT_LIST_LIMIT: usize = 1000;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct RootPathParams {
    root: String,
    path: String,
    #[serde(default)]
    max_bytes: Option<u64>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct WriteParams {
    root: String,
    path: String,
    content: String,
    #[serde(default)]
    encoding: Option<String>,
    #[serde(default)]
    max_bytes: Option<u64>,
    #[serde(default)]
    must_not_exist: Option<bool>,
    #[serde(default = "default_true")]
    create_parents: bool,
}

fn default_true() -> bool {
    true
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct DeleteParams {
    root: String,
    path: String,
    #[serde(default)]
    recursive: bool,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct TransferParams {
    root: String,
    from: String,
    to: String,
    #[serde(default)]
    overwrite: bool,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ListParams {
    root: String,
    #[serde(default)]
    path: Option<String>,
    #[serde(default)]
    glob: Option<String>,
    #[serde(default)]
    max_depth: Option<usize>,
    #[serde(default)]
    limit: Option<usize>,
    #[serde(default)]
    exclude: Vec<String>,
}

fn guard_for(root: &str) -> CoreResult<WorkspaceGuard> {
    WorkspaceGuard::new(root)
}

pub fn read(params: Value) -> CoreResult<Value> {
    let p: RootPathParams = serde_json::from_value(params)
        .map_err(|e| CoreError::invalid_argument(format!("Invalid params: {e}")))?;
    let guard = guard_for(&p.root)?;
    let resolved = guard.resolve(&p.path)?;

    let metadata = fs::metadata(&resolved).map_err(|e| match e.kind() {
        std::io::ErrorKind::NotFound => CoreError::file_not_found(&p.path),
        _ => CoreError::from(e),
    })?;
    if metadata.is_dir() {
        return Err(CoreError::is_a_directory(&p.path));
    }

    let limit = p.max_bytes.unwrap_or(DEFAULT_MAX_READ_BYTES);
    if metadata.len() > limit {
        return Err(CoreError::file_too_large(limit));
    }

    let bytes = fs::read(&resolved).map_err(CoreError::from)?;
    let (content, encoding) = match String::from_utf8(bytes.clone()) {
        Ok(text) => (text, "utf8"),
        Err(_) => (
            base64::engine::general_purpose::STANDARD.encode(&bytes),
            "base64",
        ),
    };

    Ok(json!({
        "path": guard.to_relative_display(&resolved),
        "content": content,
        "encoding": encoding,
        "size": bytes.len() as u64,
    }))
}

pub fn write(params: Value) -> CoreResult<Value> {
    let p: WriteParams = serde_json::from_value(params)
        .map_err(|e| CoreError::invalid_argument(format!("Invalid params: {e}")))?;
    let guard = guard_for(&p.root)?;
    let resolved = guard.resolve(&p.path)?;

    if p.must_not_exist.unwrap_or(false) && resolved.exists() {
        return Err(CoreError::already_exists(&p.path));
    }
    if resolved.is_dir() {
        return Err(CoreError::is_a_directory(&p.path));
    }

    let bytes: Vec<u8> = match p.encoding.as_deref() {
        Some("base64") => base64::engine::general_purpose::STANDARD
            .decode(&p.content)
            .map_err(|e| CoreError::invalid_argument(format!("Invalid base64 content: {e}")))?,
        _ => p.content.into_bytes(),
    };

    let limit = p.max_bytes.unwrap_or(DEFAULT_MAX_WRITE_BYTES);
    if bytes.len() as u64 > limit {
        return Err(CoreError::file_too_large(limit));
    }

    if let Some(parent) = resolved.parent() {
        if !parent.exists() {
            if p.create_parents {
                fs::create_dir_all(parent).map_err(CoreError::from)?;
            } else {
                return Err(CoreError::invalid_argument(format!(
                    "Parent directory does not exist: {}",
                    guard.to_relative_display(parent)
                )));
            }
        }
    }

    // Write atomically: stage in a sibling temp file, then rename over the
    // target so a crash or concurrent read never observes a partial write.
    let parent = resolved.parent().unwrap_or_else(|| guard.root());
    let mut tmp_path = parent.to_path_buf();
    let tmp_name = format!(
        ".codelink-tmp-{}-{}",
        std::process::id(),
        std::time::SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_nanos())
            .unwrap_or(0)
    );
    tmp_path.push(tmp_name);

    let mut file = fs::File::create(&tmp_path).map_err(CoreError::from)?;
    file.write_all(&bytes).map_err(CoreError::from)?;
    file.sync_all().ok();
    drop(file);
    fs::rename(&tmp_path, &resolved).map_err(|e| {
        let _ = fs::remove_file(&tmp_path);
        CoreError::from(e)
    })?;

    Ok(json!({
        "path": guard.to_relative_display(&resolved),
        "bytesWritten": bytes.len() as u64,
    }))
}

pub fn delete(params: Value) -> CoreResult<Value> {
    let p: DeleteParams = serde_json::from_value(params)
        .map_err(|e| CoreError::invalid_argument(format!("Invalid params: {e}")))?;
    let guard = guard_for(&p.root)?;
    let resolved = guard.resolve(&p.path)?;
    guard.guard_against_root_deletion(&resolved)?;

    let metadata = fs::symlink_metadata(&resolved).map_err(|e| match e.kind() {
        std::io::ErrorKind::NotFound => CoreError::file_not_found(&p.path),
        _ => CoreError::from(e),
    })?;

    if metadata.is_dir() {
        if p.recursive {
            fs::remove_dir_all(&resolved).map_err(CoreError::from)?;
        } else {
            fs::remove_dir(&resolved).map_err(CoreError::from)?;
        }
    } else {
        fs::remove_file(&resolved).map_err(CoreError::from)?;
    }

    Ok(json!({ "path": guard.to_relative_display(&resolved), "deleted": true }))
}

pub fn move_path(params: Value) -> CoreResult<Value> {
    let p: TransferParams = serde_json::from_value(params)
        .map_err(|e| CoreError::invalid_argument(format!("Invalid params: {e}")))?;
    let guard = guard_for(&p.root)?;
    let from = guard.resolve(&p.from)?;
    let to = guard.resolve(&p.to)?;
    guard.guard_against_root_deletion(&from)?;

    if !from.exists() {
        return Err(CoreError::file_not_found(&p.from));
    }
    if to.exists() && !p.overwrite {
        return Err(CoreError::already_exists(&p.to));
    }
    if let Some(parent) = to.parent() {
        fs::create_dir_all(parent).map_err(CoreError::from)?;
    }
    fs::rename(&from, &to).map_err(CoreError::from)?;

    Ok(json!({
        "from": guard.to_relative_display(&from),
        "to": guard.to_relative_display(&to),
    }))
}

pub fn copy_path(params: Value) -> CoreResult<Value> {
    let p: TransferParams = serde_json::from_value(params)
        .map_err(|e| CoreError::invalid_argument(format!("Invalid params: {e}")))?;
    let guard = guard_for(&p.root)?;
    let from = guard.resolve(&p.from)?;
    let to = guard.resolve(&p.to)?;

    if !from.exists() {
        return Err(CoreError::file_not_found(&p.from));
    }
    if to.exists() && !p.overwrite {
        return Err(CoreError::already_exists(&p.to));
    }
    if let Some(parent) = to.parent() {
        fs::create_dir_all(parent).map_err(CoreError::from)?;
    }

    if from.is_dir() {
        copy_dir_recursive(&from, &to)?;
    } else {
        fs::copy(&from, &to).map_err(CoreError::from)?;
    }

    Ok(json!({
        "from": guard.to_relative_display(&from),
        "to": guard.to_relative_display(&to),
    }))
}

fn copy_dir_recursive(from: &Path, to: &Path) -> CoreResult<()> {
    fs::create_dir_all(to).map_err(CoreError::from)?;
    for entry in fs::read_dir(from).map_err(CoreError::from)? {
        let entry = entry.map_err(CoreError::from)?;
        let file_type = entry.file_type().map_err(CoreError::from)?;
        let dest = to.join(entry.file_name());
        if file_type.is_dir() {
            copy_dir_recursive(&entry.path(), &dest)?;
        } else {
            fs::copy(entry.path(), &dest).map_err(CoreError::from)?;
        }
    }
    Ok(())
}

pub fn exists(params: Value) -> CoreResult<Value> {
    let p: RootPathParams = serde_json::from_value(params)
        .map_err(|e| CoreError::invalid_argument(format!("Invalid params: {e}")))?;
    let guard = guard_for(&p.root)?;
    let resolved = guard.resolve(&p.path)?;

    match fs::symlink_metadata(&resolved) {
        Ok(metadata) => Ok(json!({
            "exists": true,
            "isDirectory": metadata.is_dir(),
            "isFile": metadata.is_file(),
            "size": metadata.len(),
        })),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
            Ok(json!({ "exists": false, "isDirectory": false, "isFile": false, "size": 0 }))
        }
        Err(e) => Err(CoreError::from(e)),
    }
}

pub fn list(params: Value) -> CoreResult<Value> {
    let p: ListParams = serde_json::from_value(params)
        .map_err(|e| CoreError::invalid_argument(format!("Invalid params: {e}")))?;
    let guard = guard_for(&p.root)?;
    let start = guard.resolve(p.path.as_deref().unwrap_or(""))?;
    if !start.exists() {
        return Err(CoreError::file_not_found(p.path.unwrap_or_default()));
    }

    let exclude_set = build_exclude_set(&p.exclude)?;
    let glob_matcher = match &p.glob {
        Some(pattern) => Some(
            globset::Glob::new(pattern)
                .map_err(|e| CoreError::invalid_argument(format!("Invalid glob '{pattern}': {e}")))?
                .compile_matcher(),
        ),
        None => None,
    };

    let limit = p.limit.unwrap_or(DEFAULT_LIST_LIMIT);
    let mut builder = ignore::WalkBuilder::new(&start);
    builder.hidden(false).standard_filters(false);
    if let Some(depth) = p.max_depth {
        builder.max_depth(Some(depth));
    }

    // Prune excluded directories from the walk itself (not just the output):
    // an excluded directory like `node_modules` is checked as if it had a
    // child (`node_modules/probe`) so a pattern of the form `**/dir/**`
    // excludes the directory entry too, and the walker never descends into
    // it, which matters for directories that can be very large.
    let filter_guard = guard.clone();
    let filter_exclude_set = exclude_set.clone();
    let filter_start = start.clone();
    builder.filter_entry(move |entry| {
        let path = entry.path();
        if path == filter_start {
            return true;
        }
        let is_dir = entry.file_type().map(|t| t.is_dir()).unwrap_or(false);
        let relative_display = filter_guard.to_relative_display(path);
        let check_path = if is_dir {
            format!("{relative_display}/probe")
        } else {
            relative_display
        };
        !is_excluded(&check_path, &filter_exclude_set)
    });

    let mut entries = Vec::new();
    let mut truncated = false;
    for result in builder.build() {
        let entry = match result {
            Ok(entry) => entry,
            Err(_) => continue,
        };
        let path = entry.path();
        if path == start {
            continue;
        }
        let relative_display = guard.to_relative_display(path);
        if let Some(matcher) = &glob_matcher {
            if !matcher.is_match(&relative_display) {
                continue;
            }
        }

        if entries.len() >= limit {
            truncated = true;
            break;
        }

        let metadata = entry.metadata().ok();
        entries.push(FileEntry {
            path: relative_display,
            name: entry.file_name().to_string_lossy().to_string(),
            is_dir: metadata.as_ref().map(|m| m.is_dir()).unwrap_or(false),
            is_symlink: entry.path_is_symlink(),
            size: metadata.as_ref().map(|m| m.len()).unwrap_or(0),
            modified_ms: metadata
                .as_ref()
                .and_then(|m| m.modified().ok())
                .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
                .map(|d| d.as_millis()),
        });
    }

    Ok(json!({ "entries": entries, "truncated": truncated }))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn root_json(dir: &Path) -> String {
        dir.to_string_lossy().to_string()
    }

    #[test]
    fn write_then_read_roundtrip() {
        let dir = tempfile::tempdir().unwrap();
        let root = root_json(dir.path());
        write(json!({ "root": root, "path": "a.txt", "content": "hello" })).unwrap();
        let result = read(json!({ "root": root, "path": "a.txt" })).unwrap();
        assert_eq!(result["content"], "hello");
        assert_eq!(result["encoding"], "utf8");
    }

    #[test]
    fn read_rejects_oversized_file() {
        let dir = tempfile::tempdir().unwrap();
        let root = root_json(dir.path());
        write(json!({ "root": root, "path": "big.txt", "content": "0123456789" })).unwrap();
        let err = read(json!({ "root": root, "path": "big.txt", "maxBytes": 5 })).unwrap_err();
        assert_eq!(err.code, "FILE_TOO_LARGE");
    }

    #[test]
    fn write_respects_max_bytes() {
        let dir = tempfile::tempdir().unwrap();
        let root = root_json(dir.path());
        let err = write(json!({
            "root": root,
            "path": "a.txt",
            "content": "hello world",
            "maxBytes": 2
        }))
        .unwrap_err();
        assert_eq!(err.code, "FILE_TOO_LARGE");
    }

    #[test]
    fn read_rejects_traversal() {
        let dir = tempfile::tempdir().unwrap();
        let root = root_json(dir.path());
        let err = read(json!({ "root": root, "path": "../../etc/passwd" })).unwrap_err();
        assert_eq!(err.code, "PATH_OUTSIDE_WORKSPACE");
    }

    #[test]
    fn delete_refuses_workspace_root() {
        let dir = tempfile::tempdir().unwrap();
        let root = root_json(dir.path());
        let err = delete(json!({ "root": root, "path": "." })).unwrap_err();
        assert_eq!(err.code, "INVALID_ARGUMENT");
    }

    #[test]
    fn create_fails_when_must_not_exist_and_present() {
        let dir = tempfile::tempdir().unwrap();
        let root = root_json(dir.path());
        write(json!({ "root": root, "path": "a.txt", "content": "1" })).unwrap();
        let err = write(json!({
            "root": root,
            "path": "a.txt",
            "content": "2",
            "mustNotExist": true
        }))
        .unwrap_err();
        assert_eq!(err.code, "ALREADY_EXISTS");
    }

    #[test]
    fn list_respects_exclude_patterns() {
        let dir = tempfile::tempdir().unwrap();
        std::fs::create_dir_all(dir.path().join("node_modules")).unwrap();
        std::fs::write(dir.path().join("node_modules/pkg.js"), "x").unwrap();
        std::fs::write(dir.path().join("app.ts"), "x").unwrap();
        let root = root_json(dir.path());
        let result = list(json!({
            "root": root,
            "exclude": ["**/node_modules/**"]
        }))
        .unwrap();
        let entries = result["entries"].as_array().unwrap();
        assert!(entries.iter().any(|e| e["path"] == "app.ts"));
        assert!(!entries
            .iter()
            .any(|e| e["path"].as_str().unwrap().contains("node_modules")));
    }
}

File: extension/src/core/filesystem.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { CodeLinkError, ErrorCodes } from "../utils/errors.js";
import { buildExcludeMatcher, compileGlob } from "./glob.js";
import { walk } from "./walk.js";
import { WorkspaceGuard } from "./workspaceGuard.js";

const DEFAULT_MAX_READ_BYTES = 10 * 1024 * 1024;
const DEFAULT_MAX_WRITE_BYTES = 10 * 1024 * 1024;
const DEFAULT_LIST_LIMIT = 1000;

function isNotFound(error: unknown): boolean {
  return (error as NodeJS.ErrnoException)?.code === "ENOENT";
}

async function ioError(error: unknown, notFoundPath: string): Promise<never> {
  if (isNotFound(error)) {
    throw new CodeLinkError(ErrorCodes.FILE_NOT_FOUND, `File not found: ${notFoundPath}`);
  }
  const code = (error as NodeJS.ErrnoException)?.code;
  if (code === "EACCES" || code === "EPERM") {
    throw new CodeLinkError(ErrorCodes.OS_PERMISSION_DENIED, (error as Error).message);
  }
  if (code === "EEXIST") {
    throw new CodeLinkError(ErrorCodes.ALREADY_EXISTS, (error as Error).message);
  }
  throw new CodeLinkError(ErrorCodes.IO_ERROR, (error as Error).message);
}

function isValidUtf8(buffer: Buffer): boolean {
  return Buffer.from(buffer.toString("utf8"), "utf8").equals(buffer);
}

export interface ReadParams {
  root: string;
  path: string;
  maxBytes?: number;
}

export async function read(params: ReadParams): Promise<{ path: string; content: string; encoding: string; size: number }> {
  const guard = await WorkspaceGuard.create(params.root);
  const resolved = await guard.resolve(params.path);

  let stat: fs.Stats;
  try {
    stat = await fs.promises.stat(resolved);
  } catch (error) {
    return ioError(error, params.path);
  }
  if (stat.isDirectory()) {
    throw new CodeLinkError(ErrorCodes.IS_A_DIRECTORY, `Is a directory: ${params.path}`);
  }

  const limit = params.maxBytes ?? DEFAULT_MAX_READ_BYTES;
  if (stat.size > limit) {
    throw new CodeLinkError(ErrorCodes.FILE_TOO_LARGE, `File exceeds the configured limit of ${limit} bytes`);
  }

  const bytes = await fs.promises.readFile(resolved);
  const encoding = isValidUtf8(bytes) ? "utf8" : "base64";
  const content = encoding === "utf8" ? bytes.toString("utf8") : bytes.toString("base64");

  return { path: guard.toRelativeDisplay(resolved), content, encoding, size: bytes.length };
}

export interface WriteParams {
  root: string;
  path: string;
  content: string;
  encoding?: string;
  maxBytes?: number;
  mustNotExist?: boolean;
  createParents?: boolean;
}

export async function write(params: WriteParams): Promise<{ path: string; bytesWritten: number }> {
  const guard = await WorkspaceGuard.create(params.root);
  const resolved = await guard.resolve(params.path);

  if (params.mustNotExist) {
    try {
      await fs.promises.access(resolved);
      throw new CodeLinkError(ErrorCodes.ALREADY_EXISTS, `Already exists: ${params.path}`);
    } catch (error) {
      if (CodeLinkError.isCodeLinkError(error)) {
        throw error;
      }
      // ENOENT is the expected, non-error case here (file does not exist yet).
    }
  }
  try {
    const stat = await fs.promises.stat(resolved);
    if (stat.isDirectory()) {
      throw new CodeLinkError(ErrorCodes.IS_A_DIRECTORY, `Is a directory: ${params.path}`);
    }
  } catch (error) {
    if (CodeLinkError.isCodeLinkError(error)) {
      throw error;
    }
    if (!isNotFound(error)) {
      return ioError(error, params.path);
    }
  }

  const bytes = params.encoding === "base64" ? Buffer.from(params.content, "base64") : Buffer.from(params.content, "utf8");
  const limit = params.maxBytes ?? DEFAULT_MAX_WRITE_BYTES;
  if (bytes.length > limit) {
    throw new CodeLinkError(ErrorCodes.FILE_TOO_LARGE, `File exceeds the configured limit of ${limit} bytes`);
  }

  const parent = path.dirname(resolved);
  const parentExists = await fs.promises
    .access(parent)
    .then(() => true)
    .catch(() => false);
  if (!parentExists) {
    if (params.createParents ?? true) {
      await fs.promises.mkdir(parent, { recursive: true }).catch((error: unknown) => ioError(error, params.path));
    } else {
      throw new CodeLinkError(
        ErrorCodes.INVALID_ARGUMENT,
        `Parent directory does not exist: ${guard.toRelativeDisplay(parent)}`,
      );
    }
  }

  // Write atomically: stage in a sibling temp file, then rename over the
  // target so a crash or concurrent read never observes a partial write.
  const tmpPath = path.join(parent, `.codelink-tmp-${process.pid}-${Date.now()}-${Math.random().toString(36).slice(2)}`);
  try {
    await fs.promises.writeFile(tmpPath, bytes);
    await fs.promises.rename(tmpPath, resolved);
  } catch (error) {
    await fs.promises.rm(tmpPath, { force: true }).catch(() => undefined);
    return ioError(error, params.path);
  }

  return { path: guard.toRelativeDisplay(resolved), bytesWritten: bytes.length };
}

export interface DeleteParams {
  root: string;
  path: string;
  recursive?: boolean;
}

export async function deleteEntry(params: DeleteParams): Promise<{ path: string; deleted: true }> {
  const guard = await WorkspaceGuard.create(params.root);
  const resolved = await guard.resolve(params.path);
  guard.guardAgainstRootDeletion(resolved);

  let stat: fs.Stats;
  try {
    stat = await fs.promises.lstat(resolved);
  } catch (error) {
    return ioError(error, params.path);
  }

  try {
    if (stat.isDirectory()) {
      if (params.recursive) {
        await fs.promises.rm(resolved, { recursive: true, force: true });
      } else {
        await fs.promises.rmdir(resolved);
      }
    } else {
      await fs.promises.unlink(resolved);
    }
  } catch (error) {
    return ioError(error, params.path);
  }

  return { path: guard.toRelativeDisplay(resolved), deleted: true };
}

export interface TransferParams {
  root: string;
  from: string;
  to: string;
  overwrite?: boolean;
}

async function fileExists(target: string): Promise<boolean> {
  try {
    await fs.promises.access(target);
    return true;
  } catch {
    return false;
  }
}

export async function movePath(params: TransferParams): Promise<{ from: string; to: string }> {
  const guard = await WorkspaceGuard.create(params.root);
  const from = await guard.resolve(params.from);
  const to = await guard.resolve(params.to);
  guard.guardAgainstRootDeletion(from);

  if (!(await fileExists(from))) {
    throw new CodeLinkError(ErrorCodes.FILE_NOT_FOUND, `File not found: ${params.from}`);
  }
  if ((await fileExists(to)) && !params.overwrite) {
    throw new CodeLinkError(ErrorCodes.ALREADY_EXISTS, `Already exists: ${params.to}`);
  }
  await fs.promises.mkdir(path.dirname(to), { recursive: true }).catch((error: unknown) => ioError(error, params.to));
  try {
    await fs.promises.rename(from, to);
  } catch (error) {
    return ioError(error, params.from);
  }

  return { from: guard.toRelativeDisplay(from), to: guard.toRelativeDisplay(to) };
}

async function copyDirRecursive(from: string, to: string): Promise<void> {
  await fs.promises.mkdir(to, { recursive: true });
  const entries = await fs.promises.readdir(from, { withFileTypes: true });
  for (const entry of entries) {
    const fromChild = path.join(from, entry.name);
    const toChild = path.join(to, entry.name);
    if (entry.isDirectory()) {
      await copyDirRecursive(fromChild, toChild);
    } else {
      await fs.promises.copyFile(fromChild, toChild);
    }
  }
}

export async function copyPath(params: TransferParams): Promise<{ from: string; to: string }> {
  const guard = await WorkspaceGuard.create(params.root);
  const from = await guard.resolve(params.from);
  const to = await guard.resolve(params.to);

  if (!(await fileExists(from))) {
    throw new CodeLinkError(ErrorCodes.FILE_NOT_FOUND, `File not found: ${params.from}`);
  }
  if ((await fileExists(to)) && !params.overwrite) {
    throw new CodeLinkError(ErrorCodes.ALREADY_EXISTS, `Already exists: ${params.to}`);
  }
  await fs.promises.mkdir(path.dirname(to), { recursive: true }).catch((error: unknown) => ioError(error, params.to));

  try {
    const stat = await fs.promises.stat(from);
    if (stat.isDirectory()) {
      await copyDirRecursive(from, to);
    } else {
      await fs.promises.copyFile(from, to);
    }
  } catch (error) {
    return ioError(error, params.from);
  }

  return { from: guard.toRelativeDisplay(from), to: guard.toRelativeDisplay(to) };
}

export interface ExistsParams {
  root: string;
  path: string;
}

export async function exists(
  params: ExistsParams,
): Promise<{ exists: boolean; isDirectory: boolean; isFile: boolean; size: number }> {
  const guard = await WorkspaceGuard.create(params.root);
  const resolved = await guard.resolve(params.path);

  try {
    const stat = await fs.promises.lstat(resolved);
    return { exists: true, isDirectory: stat.isDirectory(), isFile: stat.isFile(), size: stat.size };
  } catch (error) {
    if (isNotFound(error)) {
      return { exists: false, isDirectory: false, isFile: false, size: 0 };
    }
    return ioError(error, params.path);
  }
}

export interface ListParams {
  root: string;
  path?: string;
  glob?: string;
  maxDepth?: number;
  limit?: number;
  exclude?: string[];
}

export interface FileEntry {
  path: string;
  name: string;
  isDir: boolean;
  isSymlink: boolean;
  size: number;
  modifiedMs?: number;
}

export async function list(params: ListParams): Promise<{ entries: FileEntry[]; truncated: boolean }> {
  const guard = await WorkspaceGuard.create(params.root);
  const start = await guard.resolve(params.path ?? "");
  if (!(await fileExists(start))) {
    throw new CodeLinkError(ErrorCodes.FILE_NOT_FOUND, `File not found: ${params.path ?? ""}`);
  }

  const isExcluded = buildExcludeMatcher(params.exclude ?? []);
  const globMatcher = params.glob ? compileGlob(params.glob) : undefined;
  const limit = params.limit ?? DEFAULT_LIST_LIMIT;

  const entries: FileEntry[] = [];
  let truncated = false;

  for await (const entry of walk(guard, start, {
    maxDepth: params.maxDepth,
    isExcluded: (probePath) => isExcluded(probePath),
  })) {
    if (globMatcher && !globMatcher.test(entry.relativeDisplay)) {
      continue;
    }
    if (entries.length >= limit) {
      truncated = true;
      break;
    }
    entries.push({
      path: entry.relativeDisplay,
      name: entry.name,
      isDir: entry.isDir,
      isSymlink: entry.isSymlink,
      size: entry.stat.size,
      modifiedMs: Math.floor(entry.stat.mtimeMs),
    });
  }

  return { entries, truncated };
}
